tools/build.cmd

```diff
@@ -157,6 +157,21 @@ rem ----------------------------------------------------------------------------
 rem this is setting the cmake path in a quoted way
 set "cmake-root=%build-root%\build"
 
+rem Figure out which CMake generator to use
+if "%VisualStudioVersion%" == "14.0" (
+    set "cmake-generator=Visual Studio 14 2015"
+) else if "%VisualStudioVersion%" == "15.0" (
+    set "cmake-generator=Visual Studio 15 2017"
+)
+if NOT DEFINED cmake-generator (
+    echo ERROR: VisualStudioVersion not defined, or not one of ^(14.0, 15.0^)
+    exit /b 1
+)
+if "%build-platform%" == "x64" (
+    set "cmake-generator=!cmake-generator! Win64"
```

Comment:
Actually, the ! syntax isn't really necessary here; I'll change it.

You need to use the ! syntax when you **set** and then **use** a variable **inside an `if` statement**. E.g.
```
@setlocal
@echo off
set hello=world
echo %hello%
if "%hello%" == "world" (
    set "hello=cruel %hello%"
    echo %hello%
)
echo %hello%
endlocal
```
...gives this output:
```
world
world
cruel world
```
Oops! The 2nd line should say "cruel world", but since the entire if-block was evaluated before it was executed, `%hello%` resolved to its pre-if-block value. But change it to this:
```
@setlocal EnableDelayedExpansion
@echo off
set hello=world
echo %hello%
if "%hello%" == "world" (
    set "hello=cruel %hello%"
    echo !hello!
)
echo %hello%
endlocal
```
...and you get the expected output:
```
world
cruel world
cruel world
```
Expansion of the variable inside the !s is delayed until the line is actually executed, so it gets the right value.